examples_tests/31.SSBOBenchmark/main.cpp

```diff
@@ -0,0 +1,401 @@
+        auxCtx->setActiveUBO(_bnd, 1, &glbuf, &_off, &_sz);
+    }
+
+    video::IGPUBuffer* ubo;
+
+private:
+    video::IVideoDriver* drv;
+    video::IGPUBuffer* mappedBuf;
+    uint8_t* mappedMem;
+    uint8_t updateNum;
+    core::smart_refctd_ptr<video::IDriverFence> fence[4];
+};*/
+
+//#define BENCH
+
+#include "irr/irrpack.h"
+struct Vertex
+{
+    uint32_t boneID;
+    float pos[3];
+    uint8_t color[4];
+    uint8_t uv[2];
+    float normal[3];
+} PACK_STRUCT;
+#include "irr/irrunpack.h"
+
+#include "common.glsl"
+#include "commonIndirect.glsl"
+
+        //TODO: create shader, and then from check how introspector creates renderpass
+
+int main()
+{
+    // create device with full flexibility over creation parameters
+    // you can add more parameters if desired, check irr::SIrrlichtCreationParameters
+    irr::SIrrlichtCreationParameters params;
+    params.Bits = 24; //may have to set to 32bit for some platforms
+    params.ZBufferBits = 24; //we'd like 32bit here
+    params.DriverType = video::EDT_OPENGL; //! Only Well functioning driver, software renderer left for sake of 2D image drawing
+    params.WindowSize = dimension2d<uint32_t>(1280, 720);
+    params.Fullscreen = false;
+    params.Vsync = false;
+    params.Doublebuffer = true;
+    params.Stencilbuffer = false; //! This will not even be a choice soon
+    auto device = createDeviceEx(params);
+
+    if (!device)
+        return 1; // could not create selected driver.
+
+    QToQuitEventReceiver receiver;
+    device->setEventReceiver(&receiver);
+
+    auto* am = device->getAssetManager();
+    video::IVideoDriver* driver = device->getVideoDriver();
+
+    IAssetLoader::SAssetLoadParams lp;
+
+    auto vertexShaderBundle = am->getAsset("../ssboBenchmarkShaders/vertexShader.vert", lp);
+    auto fragmentShaderBundle = am->getAsset("../ssboBenchmarkShaders/fragmentShader.frag", lp);
+
+    CShaderIntrospector introspector(am->getGLSLCompiler());
+    const auto extensions = driver->getSupportedGLSLExtensions();
+
+    //create cpu shaders
+    ICPUSpecializedShader* shaders[2] =
+    {
+        IAsset::castDown<ICPUSpecializedShader>(vertexShaderBundle.getContents().first->get()),
+        IAsset::castDown<ICPUSpecializedShader>(fragmentShaderBundle.getContents().first->get())
+    };
+    auto cpuPipeline = introspector.createApproximateRenderpassIndependentPipelineFromIntrospection(shaders, shaders + 2, extensions->begin(), extensions->end());
+
+    //temporary
+    constexpr size_t diskCount = 3001;
+
+    std::vector<uint16_t> tesselation(diskCount);
+
+    //get random tesselation for disks
+    {
+        std::random_device rd;
+        std::mt19937 mt(rd());
+        std::uniform_int_distribution<uint32_t> dist(15, 14999);
+
+        //TODO: test
+        //`dist(mt) | 0x0001` so vertexCount is always odd (only when diskCount is odd as well)
+        std::generate(tesselation.begin(), tesselation.end(), [&]() { return dist(mt) | 0x0001; });
```

Comment:
formula for vertex count in a disk 1+triangleCount, which means that the number of triangles should be even to have odd numbers of vertices (and hence vertex shader invocations)

Reply:
its 2+triangleCount if there's no vertex re-use of the vertex creating the seam at the circumference